slideflow/slide/wsi.py

```diff
@@ -446,32 +452,97 @@ def _build_coord(self) -> None:
             x_offset = - (full_extract/2 - stride/2)
             y_offset = - (full_extract/2 - stride/2)
 
-            # Translate ROI polygons
-            translated = [
-                sa.translate(roi.poly, x_offset, y_offset)
-                for roi in self.rois
-            ]
-
-            # Set scale to 50 times greater than grid size
-            # if filtering by float
-            o = 1 if roi_by_center else 50
-
-            # Scale ROI polygons
-            scaled = [
-                sa.scale(poly, xfact=xfact * o, yfact=yfact * o, origin=(0, 0))
-                for poly in translated
-            ]
-
-            # Rasterize polygons to the size of the tile extraction grid.
-            # Rasterize polygons for ROIs individually, to keep track of
-            # which ROI each tile belongs to, then merge.
-            self.roi_grid = np.stack([
-                rasterio.features.rasterize(
-                    [scaled_roi],
-                    out_shape=(self.grid.shape[1] * o, self.grid.shape[0] * o),
-                    all_touched=False).astype(bool).astype(int) * (i + 1)
-                for i, scaled_roi in enumerate(scaled)
-            ], axis=0).max(axis=0).T
+            if self.artifact_rois is None:
```

Comment:
Let's make this part less redudant. 

Change self.artifact_rois from None to empty list so that we can always run 

```
translated = [
    sa.translate(roi.poly, x_offset, y_offset)
    for roi in self.rois if roi.label not in self.artifact_rois
]
```

and all the code below. 

Then if self.artifact_rois is not a empyt list we compute translated_issues, scaled_issues, roi_grid_issues

And finally if there weren't any artifact rois

self.roi_grid is unchanged

if there were only artifact_rois 

self.roi_grid = roi_grid_issues.T

if both were there 

self.roi_grid = = np.minimum(roi_grid_tissues, self.roi_grid).T

NB: no more stacking and .min, use single operation np.minimum

Reply:
Finish updating this.